src/exp/istio-watcher/pkg/watcher/watcher.go

```diff
@@ -0,0 +1,290 @@
+)
+
+var (
+	ConnectionInfoInsufficient = errors.New("connection info partial or empty")
+)
+
+type IstioWatcher struct {
+	clientset    *kubernetes.Clientset
+	config       *rest.Config
+	mapperClient mapperclient.MapperClient
+	connections  map[*ConnectionWithPath]time.Time
+}
+
+type ConnectionWithPath struct {
+	SourceWorkload       string `regroup:"source_workload"`
+	SourceNamespace      string `regroup:"source_namespace"`
+	DestinationWorkload  string `regroup:"destination_workload"`
+	DestinationNamespace string `regroup:"destination_namespace"`
+	RequestPath          string `regroup:"request_path"`
+}
+
+func (p *ConnectionWithPath) hasMissingInfo() bool {
+	for _, field := range []string{p.SourceWorkload, p.SourceNamespace, p.DestinationWorkload, p.DestinationNamespace} {
+		if field == "" || strings.Contains(field, "unknown") {
+			return true
+		}
+	}
+
+	return p.RequestPath == ""
+}
+
+// omitMetricsFieldsFromConnection drops the metric name and uses the value alone in the connection
+// Since we cant use lookaheads in our regex matching, connections fields are parsed with their metric name as well
+// e.g. for source workload we get "source_workload.some-client", and we need to parse "some-client" and remove the metric name
+func (p *ConnectionWithPath) omitMetricsFieldsFromConnection() {
+	p.SourceWorkload = strings.Split(p.SourceWorkload, ".")[1]
+	p.DestinationWorkload = strings.Split(p.DestinationWorkload, ".")[1]
+	p.SourceNamespace = strings.Split(p.SourceNamespace, ".")[1]
+	p.DestinationNamespace = strings.Split(p.DestinationNamespace, ".")[1]
+	p.RequestPath = strings.Split(p.RequestPath, ".")[1]
+}
+
+type EnvoyMetrics struct {
+	Stats []Metric `json:"stats"`
+}
+
+type Metric struct {
+	Name string `json:"name"`
+}
+
+func NewWatcher(mapperClient mapperclient.MapperClient) (*IstioWatcher, error) {
+	conf, err := rest.InClusterConfig()
+
+	if err != nil && !errors.Is(err, rest.ErrNotInCluster) {
+		return nil, err
+	}
+
+	// We try building the REST Config from ./kube/config to support running the watcher locally
+	if conf == nil {
+		conf, err = clientcmd.BuildConfigFromFlags("", filepath.Join(homedir.HomeDir(), ".kube", "config"))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	clientset, err := kubernetes.NewForConfig(conf)
+	if err != nil {
+		return nil, err
+	}
+
+	m := &IstioWatcher{
+		clientset:    clientset,
+		config:       conf,
+		mapperClient: mapperClient,
+		connections:  map[*ConnectionWithPath]time.Time{},
+	}
+
+	return m, nil
+}
+
+func (m *IstioWatcher) Flush() map[*ConnectionWithPath]time.Time {
+	r := m.connections
+	m.connections = map[*ConnectionWithPath]time.Time{}
+	return r
+}
+
+func (m *IstioWatcher) CollectIstioConnectionMetrics(ctx context.Context, namespace string) error {
+	sendersErrGroup, sendersCtx := errgroup.WithContext(ctx)
```

Comment:
Use `SetLimit` to limit concurrency to some small number, like 5 or 10, otherwise it may run a lot of goroutines all at once for particularly big namespaces.